packages/service-library/src/servicelib/archiving_utils.py

```diff
@@ -168,20 +183,39 @@ async def unarchive_dir(
                 for zip_entry in zip_file_handler.infolist()
             ]
 
-            extracted_paths: list[Path] = await tqdm_asyncio.gather(
-                *tasks,
-                desc=f"decompressing {archive_to_extract} -> {destination_folder} [{len(tasks)} file{'s' if len(tasks) > 1 else ''}"
-                f"/{_human_readable_size(archive_to_extract.stat().st_size)}]\n",
-                total=len(tasks),
-                **_TQDM_MULTI_FILES_OPTIONS,
-            )
+            try:
+                extracted_paths: list[Path] = await tqdm_asyncio.gather(
+                    *futures,
+                    desc=f"decompressing {archive_to_extract} -> {destination_folder} [{len(futures)} file{'s' if len(futures) > 1 else ''}"
+                    f"/{_human_readable_size(archive_to_extract.stat().st_size)}]\n",
+                    total=len(futures),
+                    **_TQDM_MULTI_FILES_OPTIONS,
+                )
+
+            except Exception as err:
+                for f in futures:
+                    f.cancel()
+
+                # wait until all tasks are cancelled.
+                await asyncio.gather(*futures, return_exceptions=True)
+
+                # now we can cleanup
+                if destination_folder.exists() and destination_folder.is_dir():
+                    shutil.rmtree(destination_folder, ignore_errors=True)
```

Comment:
Could you please use wraps like [here](https://github.com/Tinche/aiofiles/blob/d794675bfe5bb5c0acf17ffe60997ad4e821e9c2/src/aiofiles/os.py#L29)  for `shutil.rmtree` so it can be awaited?

Reply:
we already have an async version of that in ``serviceslib.file_utils``. Will use that.